### ISSUE TITLE
Required initializer is marked as not required by CA1805

### ISSUE THREAD

**Issue by AshleighAdams** [dotnet/roslyn-analyzers#5887]
### Analyzer

**Diagnostic ID**: [CA1805 ](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA1805)

### Describe the improvement

As it stands with the new maintenance release of .NET 6, if you use a `record struct` with default initializers on auto properties, all fields must be assigned to in the default constructor. This leads to an issue where the ctor forces the default value to be set, but the analyzer reports it as not required.

### Describe suggestions on how to achieve the rule

This rule should be disabled on record structs, assuming the compiler is correct in this case in that all fields must be assigned to.

### Additional context

Minimal repo:

```csharp
public record struct MyRecord
{
	public MyRecord() { }
	public bool SomeBool { get; set; } = false;
}
```

```
$ dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.200
 Commit:    4c30de7899

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.200\

Host (useful for support):
  Version: 6.0.2
  Commit:  839cdfb0ec

.NET SDKs installed:
  5.0.404 [C:\Program Files\dotnet\sdk]
  6.0.200 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```

**Comment by mavasani** [comment]
@stephentoub

**Comment by Youssef1313** [comment]
Is it really specific to `record struct`s? or `struct`s in general?

**Comment by Youssef1313** [comment]
Feels like a duplicate of https://github.com/dotnet/roslyn-analyzers/issues/5750 which is being fixed in https://github.com/dotnet/roslyn-analyzers/pull/5813

**Comment by AshleighAdams** [comment]
@Youssef1313, That looks right, my bad. I'll leave it to you if you want to close as a duplicate, or keep it open to track and ensure the fix works for both.

**Comment by stephentoub** [comment]
> Is it really specific to record structs? or structs in general?
> Feels like a duplicate of https://github.com/dotnet/roslyn-analyzers/issues/5750 which is being fixed in https://github.com/dotnet/roslyn-analyzers/pull/5813

It's structs in general, but I don't think it's a duplicate.  https://github.com/dotnet/roslyn-analyzers/issues/5750 is about a case like this:
```C#
struct S
{
    public S() { ... }
}

static S s_value = new S();
```
where `new S()` and `default(S)` are not the same because `S` has a user-defined parameterless ctor, and that makes it a functional change to replace `static S s_value = new S()` with `static S s_value`.

This is instead about this:
```C#
struct S
{
    public S() { ... }
    public int Prop { get; set; } = 42;
}
```
where S's ctor is responsible for fully initializing the whole struct, including the backing fields for auto-props, and it could be relying on that `= 42` to perform the initialization... if `S()` doesn't contain duplicative code that also does `Prop = whatever;`, then removing the `= 42;` is a breaking change that will cause this to fail to compile.

I think the answer is simply that CA1805 should never apply to structs, of any kind.

**Comment by Evangelink** [comment]
Just or me to be 100% sure what I am supposed to end up doing:

```cs
public record struct MyRecord
{
     public bool SomeBool { get; set; } = false; // Reports because there is no parameterless ctor | Same behavior if only struct
}

public record struct MyRecord2()
{
     public bool SomeBool { get; set; } = false; // Does not report - parameterless ctor
}

public record struct MyRecord3
{
    public MyRecord() { }
    public bool SomeBool { get; set; } = false;  // Does not report - parameterless ctor | Same behavior if only struct
}
```

**Comment by stephentoub** [comment]
My suggestion was that CA1805 wouldn't fire for any of your examples.  It shouldn't apply to any instance members on any struct.  SomeBool in all of these cases is an instance member on a struct. (And I believe your first example won't even compile anymore because C# now requires an explicit ctor exist in order to use field initializers on structs).
